src/main/java/de/hysky/skyblocker/skyblock/garden/VisitorHelper.java

```diff
@@ -50,55 +47,134 @@ public class VisitorHelper {
+					screen instanceof HandledScreen<?> handledScreen &&
+					(!title.contains("Logbook") || title.startsWith("Bazaar"))) {
+
+				ScreenEvents.afterRender(screen).register((screen_, context, mouseX, mouseY, delta) ->
+						renderScreen(title, context, client.textRenderer, handledScreen.getScreenHandler(), mouseX, mouseY));
+
+				ScreenEvents.remove(screen).register(screen_ -> shouldProcessVisitorItems = true);
+			}
+		});
+	}
 
     public static void renderScreen(String title, DrawContext context, TextRenderer textRenderer, ScreenHandler handler, int mouseX, int mouseY) {
         if (handler.getCursorStack() == ItemStack.EMPTY && shouldProcessVisitorItems) processVisitorItem(title, handler);
         drawScreenItems(context, textRenderer, mouseX, mouseY);
     }
 
-    public static void onMouseClicked(double mouseX, double mouseY, int mouseButton, TextRenderer textRenderer) {
-        int yPosition = TEXT_START_Y;
-        for (Map.Entry<Pair<String, String>, Object2IntMap<String>> visitorEntry : itemMap.entrySet()) {
-            yPosition += LINE_SPACING + textRenderer.fontHeight;
-
-            for (Object2IntMap.Entry<String> itemEntry : visitorEntry.getValue().object2IntEntrySet()) {
-                String itemText = itemEntry.getKey();
-                int textWidth = textRenderer.getWidth(itemText + " x" + itemEntry.getIntValue());
-
-                // Check if the mouse is over the item text
-                // The text starts at `TEXT_START_X + ENTRY_INDENT + ITEM_INDENT`
-                if (isMouseOverText(mouseX, mouseY, TEXT_START_X + ENTRY_INDENT + ITEM_INDENT, yPosition, textWidth, textRenderer.fontHeight)) {
-                    // Send command to buy the item from the bazaar
-                    MessageScheduler.INSTANCE.sendMessageAfterCooldown("/bz " + itemText, true);
-                    return;
-                }
-
-                // Check if the mouse is over the copy amount text
-                // The copy amount text starts at `TEXT_START_X + ENTRY_INDENT + ITEM_INDENT + textWidth`
-                MinecraftClient client = MinecraftClient.getInstance();
-                if (client.player != null && isMouseOverText(mouseX, mouseY, TEXT_START_X + ENTRY_INDENT + ITEM_INDENT + textWidth, yPosition, textRenderer.getWidth(" [Copy Amount]"), textRenderer.fontHeight)) {
-                    // Copy the amount to the clipboard
-                    client.keyboard.setClipboard(String.valueOf(itemEntry.getIntValue()));
-                    client.player.sendMessage(Constants.PREFIX.get().append("Copied amount successfully"), false);
-                    return;
-                }
-
-                yPosition += LINE_SPACING + textRenderer.fontHeight;
-            }
-        }
-    }
+	// The location of copy amount, and item text seem to be overlapping so clicking on part of the itemText will run copy amount.
+	public static void onMouseClicked(double mouseX, double mouseY, int mouseButton, TextRenderer textRenderer) {
```

Comment:
Definitely not a fan of the logic as is. I have a couple of suggestions:
1. You could use `SortedMap<String, Object2IntMap<Pair<String, String>>>` for `itemMap` instead, which won't keep one visitor's items together, but they usually only request one item anyways.
2. You could replace the mouse click logic with `ClickEvent` and `Screen#handleTextClick`. See how `ChatScreen#mouseClicked` does it. This way, you only have the rendering logic, and not the mouse clicked logic.